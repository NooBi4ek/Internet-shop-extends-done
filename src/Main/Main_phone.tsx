import { useDispatch, useSelector } from 'react-redux';
import React, { useEffect } from 'react';
import {
  addToOrder,
  startPhone,
  afterAddOrder,
} from '../store/internet_shop_slice.ts';
import {
  Button_buy,
  Main_onephone_block,
  Main_onephone_img,
  Main_phone_wrapper,
} from '../styled/styled-mainphone/SMainphone';
import '../styled/styled-mainphone/SMainphone.css';
import { IPhones } from '../models/models';
import { useNavigate } from 'react-router-dom';
const Main_phone = () => {
  const dispatch = useDispatch();
  const filterphone = useSelector((state) => state.shop.filter_phone);
  const navigate = useNavigate();
  useEffect(() => {
    dispatch(startPhone());
  }, []);
  return (
    <Main_phone_wrapper>
      {filterphone.map((phone: IPhones) => (
        <Main_onephone_block key={phone.id}>
          <Main_onephone_img
            src={'./img/' + phone.img}
            alt={phone.name}
            onClick={() => {
              navigate(`${phone.id}`);
            }}
          />
          <div className="phoneName">{phone.name}</div>
          <div className="phoneName">{phone.price} UAH</div>
          <Button_buy
            key={phone.id}
            onClick={() => {
              dispatch(addToOrder({ phone }));
              dispatch(afterAddOrder({ phone }));
            }}
          >
            {phone.click ? 'In the box' : 'Buy'}
          </Button_buy>
        </Main_onephone_block>
      ))}
    </Main_phone_wrapper>
  );
};
export default Main_phone;
