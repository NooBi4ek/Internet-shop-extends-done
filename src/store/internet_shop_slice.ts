import { createSlice } from '@reduxjs/toolkit';
import { IPhones } from '../models/models';

const shopSlice: IPhones = createSlice({
  name: 'shop',
  initialState: {
    phones: [
      {
        id: 1,
        name: 'Apple iPhone 11 128GB',
        company: 'Apple',
        short_desc: 'something',
        img: 'Apple_Iphone11_128.jpg',
        threads_name: 'Apple A13 Bionic',
        front_camera: '12MP',
        main_camera: '12MP',
        count: 1,
        quantity_threads: 6,
        memory: 128,
        price: 10000,
        click: false,
        click_versus: false,
      },
      {
        id: 2,
        name: 'Apple iPhone 13 128GB',
        company: 'Apple',
        short_desc: 'something',
        img: 'Apple_Iphone13_128.jpg',
        threads_name: 'Apple A14 Bionic',
        front_camera: '12MP',
        main_camera: '12MP',
        price: 15000,
        count: 1,
        quantity_threads: 6,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 3,
        name: 'Apple iPhone 13 PRO 128GB',
        company: 'Apple',
        short_desc: 'something',
        img: 'Apple_Iphone13_pro_128.jpg',
        threads_name: 'Apple A15 Bionic',
        front_camera: '12MP',
        main_camera: '12MP',
        price: 17000,
        count: 1,
        quantity_threads: 6,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 4,
        name: 'Apple iPhone 13 PRO MAX 128GB',
        company: 'Apple',
        short_desc: 'something',
        img: 'Apple_Iphone13_pro_max_128.jpg',
        threads_name: 'Apple A15 Bionic',
        front_camera: '12MP',
        main_camera: '12MP',
        price: 20000,
        count: 1,
        quantity_threads: 6,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 5,
        name: 'Infinix Note 10 PRO 8/128GB',
        company: 'Infinix',
        short_desc: 'something',
        img: 'Infinix_note10_pro_8_128.jpg',
        threads_name: 'MediaTek Helio G95',
        main_camera: '64MP + 8MP + 2MP + 2MP',
        front_camera: '16MP',
        price: 7000,
        count: 1,
        quantity_threads: 8,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 6,
        name: 'Motorola Moto G60 6/128GB',
        company: 'Motorola',
        short_desc: 'something',
        img: 'Motorola_moto_g60_6_128.jpg',
        threads_name: 'Qualcomm Snapdragon 732G',
        main_camera: '108MP + 8MP + 2MP',
        front_camera: '32MP',
        price: 9000,
        count: 1,
        quantity_threads: 8,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 7,
        name: 'Samsung Galaxy A32 4/64GB',
        company: 'Samsung',
        short_desc: 'something',
        img: 'Samsung_galaxy_a32_4_64.jpg',
        threads_name: 'MediaTek Helio G80',
        main_camera: '64MP + 8MP + 5MP + 5MP',
        front_camera: '20MP',
        price: 5000,
        count: 1,
        quantity_threads: 8,
        memory: 64,
        click: false,
        click_versus: false,
      },
      {
        id: 8,
        name: 'Samsung Galaxy A32 4/128GB',
        company: 'Samsung',
        short_desc: 'something',
        img: 'Samsung_galaxy_a32_4_128.jpg',
        threads_name: 'MediaTek Helio G80',
        main_camera: '64MP + 8MP + 5MP + 5MP',
        front_camera: '20MP',
        price: 10000,
        count: 1,
        quantity_threads: 8,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 9,
        name: 'Samsung Galaxy A53 5G 6/128GB',
        company: 'Samsung',
        short_desc: 'something',
        img: 'Samsung_galaxy_a53_5g_6_128.jpg',
        threads_name: 'Samsung Exynos 1280',
        main_camera: '64MP + 12MP + 5MP + 5MP',
        front_camera: '32MP',
        price: 17000,
        count: 1,
        quantity_threads: 8,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 10,
        name: 'Samsung Galaxy M32 6/128GB',
        company: 'Samsung',
        short_desc: 'something',
        img: 'Samsung_galaxy_m32_6_128.jpg',
        threads_name: 'MediaTek Helio G80',
        main_camera: '64MP + 8MP + 2MP + 2MP',
        front_camera: '20MP',
        price: 17000,
        count: 1,
        quantity_threads: 8,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 11,
        name: 'Samsung Galaxy M52 5G 6/128GB',
        company: 'Samsung',
        short_desc: 'something',
        img: 'Samsung_galaxy_m52_5g_6_128.jpg',
        threads_name: 'Qualcomm Snapdragon 778G',
        main_camera: '64MP + 12MP + 5MP',
        front_camera: '32MP',
        price: 11000,
        count: 1,
        quantity_threads: 8,
        memory: 128,
        click: false,
        click_versus: false,
      },
      {
        id: 12,
        name: 'Samsung Galaxy M53 5G 6/128GB',
        company: 'Samsung',
        short_desc: 'something',
        img: 'Samsung_galaxy_m53_5g_6_128.jpg',
        threads_name: 'Qualcomm Snapdragon 778G',
        main_camera: '64MP + 12MP + 5MP',
        front_camera: '32MP',
        price: 17000,
        count: 1,
        quantity_threads: 8,
        memory: 128,
        click: false,
        click_versus: false,
      },
    ],
    categories: [
      { id: 1, namecat: 'All' },
      { id: 2, namecat: 'Apple' },
      { id: 3, namecat: 'Samsung' },
      { id: 4, namecat: 'Infinix' },
    ],
    All_category: false,
    orders: [],
    filter_phone: [],
    versus_Phone: [],
    sum: 0,
  },
  reducers: {
    filterCategories(state, action) {
      state.filter_phone = [];
      if (action.payload.el.namecat === 'All') {
        state.phones.map((el) => state.filter_phone.push(el));
        state.All_category = true;
      } else {
        state.phones.filter((phones) => {
          return (
            phones.company === action.payload.el.namecat &&
            state.filter_phone.push(phones)
          );
        });
        state.All_category = false;
      }
    },
    startPhone(state) {
      state.filter_phone = [];
      state.phones.map((el) => state.filter_phone.push(el));
      state.All_category = true;
    },
    Countsum(state) {
      state.orders.forEach((el) => (state.sum += el.count * el.price));
    },
    addToOrder(state, action) {
      let isArr = false;
      state.orders.forEach((el) => {
        if (el.id === action.payload.phone.id) {
          el.count = el.count + 1;
          isArr = true;
        }
      });
      if (!isArr) {
        state.filter_phone = [];
        state.orders.push(action.payload.phone);
        state.filter_phone = state.phones.map((el) =>
          el.id === action.payload.phone.id
            ? { ...el, click: (el.click = true) }
            : el,
        );
      }
    },
    afterAddOrder(state, action) {
      if (state.All_category !== true) {
        state.filter_phone = [];
        state.phones.forEach((el) =>
          el.company === action.payload.phone.company
            ? state.filter_phone.push(el)
            : null,
        );
      } else {
        state.filter_phone = [];
        state.phones.forEach((el) => state.filter_phone.push(el));
      }
    },
    deleteOrder(state, action) {
      if (action.payload.orders.count > 1) {
        state.orders = state.orders.map((el) =>
          el.id === action.payload.orders.id
            ? { ...el, count: el.count - 1 }
            : el,
        );
      } else {
        state.filter_phone = state.phones.map((el) =>
          el.id === action.payload.orders.id
            ? { ...el, click: (el.click = false) }
            : el,
        );
        if (state.All_category !== true) {
          state.filter_phone = [];
          state.phones.filter((el) =>
            el.company === action.payload.orders.company
              ? state.filter_phone.push(el)
              : el,
          );
        } else {
          state.filter_phone = [];
          state.phones.forEach((el) => state.filter_phone.push(el));
        }
        state.orders = state.orders.filter(
          (el) => el.id !== action.payload.orders.id,
        );
      }
    },
    addToVersus(state, action) {
      let isArr = false;
      state.versus_Phone.forEach((el) => {
        if (el.id === action.payload.id) {
          isArr = true;
        }
      });
      if (!isArr) {
        state.versus_Phone.push(action.payload.phone);
        state.filter_phone = state.phones.map((el) =>
          el.id === action.payload.phone.id
            ? { ...el, click_versus: (el.click_versus = true) }
            : el,
        );
      }
    },
    DeleteInVersus(state, action) {
      state.versus_Phone = state.versus_Phone.filter(
        (el) => el.id !== action.payload.phone.id,
      );
      state.phones = state.phones.map((el) =>
        el.id === action.payload.phone.id
          ? { ...el, click_versus: (el.click_versus = false) }
          : el,
      );
    },
    addcount(state, action) {
      state.orders = state.orders.map((el) =>
        el.id === action.payload.id ? { ...el, count: el.count + 1 } : el,
      );
    },
  },
});
export const {
  addToOrder,
  addcount,
  addToVersus,
  DeleteInVersus,
  deleteOrder,
  Countsum,
  filterCategories,
  startPhone,
  afterAddOrder,
} = shopSlice.actions;
export default shopSlice.reducer;
