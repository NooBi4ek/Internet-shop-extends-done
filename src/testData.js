export const data = [
  {
    id: 1,
    name: 'Apple iPhone 11 128GB',
    company: 'Apple',
    short_desc: 'something',
    img: 'Apple_Iphone11_128.jpg',
    threads_name: 'Apple A13 Bionic',
    front_camera: '12MP',
    main_camera: '12MP',
    count: 1,
    quantity_threads: 6,
    memory: 128,
    price: 10000,
    click: false,
    click_versus: false,
  },
  {
    id: 2,
    name: 'Apple iPhone 13 128GB',
    company: 'Apple',
    short_desc: 'something',
    img: 'Apple_Iphone13_128.jpg',
    threads_name: 'Apple A14 Bionic',
    front_camera: '12MP',
    main_camera: '12MP',
    price: 15000,
    count: 1,
    quantity_threads: 6,
    memory: 128,
    click: false,
    click_versus: false,
  },
];
