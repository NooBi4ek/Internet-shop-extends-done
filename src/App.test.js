import Main_phone from './Main/Main_phone';
import * as reduxHooks from 'react-redux';
import * as action from './store/internet_shop_slice.ts';
import { data } from './testData';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
jest.mock('react-redux');
jest.mock('react-router-dom');
const mockDispatch = jest.spyOn(reduxHooks, 'useDispatch');
const mockuseSelector = jest.spyOn(reduxHooks, 'useSelector');

describe('Main Phone test', () => {
  it('should button buy is work', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValue(dispatch);
    mockuseSelector.mockReturnValue(data);
    const mockAddToOrder = jest.spyOn(action, 'addToOrder');
    const component = render(<Main_phone />);
    expect(mockAddToOrder).toBeCalledTimes(0);
    userEvent.click(screen.getAllByRole('button')[0]);
    expect(mockAddToOrder).toBeCalledTimes(1);
    expect(component).toMatchSnapshot();
  });
});
