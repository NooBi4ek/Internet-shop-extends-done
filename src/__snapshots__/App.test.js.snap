// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Phone test should button buy is work 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bcXHqe"
      >
        <div
          class="sc-gswNZR dAFeWa"
        >
          <img
            alt="Apple iPhone 11 128GB"
            class="sc-dkrFOg kaJIuH"
            src="./img/Apple_Iphone11_128.jpg"
          />
          <div
            class="phoneName"
          >
            Apple iPhone 11 128GB
          </div>
          <div
            class="phoneName"
          >
            10000
             UAH
          </div>
          <button
            class="sc-hLBbgP hXuYbx"
            title="Buy"
          >
            Buy
          </button>
        </div>
        <div
          class="sc-gswNZR dAFeWa"
        >
          <img
            alt="Apple iPhone 13 128GB"
            class="sc-dkrFOg kaJIuH"
            src="./img/Apple_Iphone13_128.jpg"
          />
          <div
            class="phoneName"
          >
            Apple iPhone 13 128GB
          </div>
          <div
            class="phoneName"
          >
            15000
             UAH
          </div>
          <button
            class="sc-hLBbgP hXuYbx"
            title="Buy"
          >
            Buy
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bcXHqe"
    >
      <div
        class="sc-gswNZR dAFeWa"
      >
        <img
          alt="Apple iPhone 11 128GB"
          class="sc-dkrFOg kaJIuH"
          src="./img/Apple_Iphone11_128.jpg"
        />
        <div
          class="phoneName"
        >
          Apple iPhone 11 128GB
        </div>
        <div
          class="phoneName"
        >
          10000
           UAH
        </div>
        <button
          class="sc-hLBbgP hXuYbx"
          title="Buy"
        >
          Buy
        </button>
      </div>
      <div
        class="sc-gswNZR dAFeWa"
      >
        <img
          alt="Apple iPhone 13 128GB"
          class="sc-dkrFOg kaJIuH"
          src="./img/Apple_Iphone13_128.jpg"
        />
        <div
          class="phoneName"
        >
          Apple iPhone 13 128GB
        </div>
        <div
          class="phoneName"
        >
          15000
           UAH
        </div>
        <button
          class="sc-hLBbgP hXuYbx"
          title="Buy"
        >
          Buy
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
